cmd/ipfs/pinmfs.go

```diff
@@ -0,0 +1,251 @@
+package main
+
+import (
+	"context"
+	_ "expvar"
+	"fmt"
+	"time"
+
+	cid "github.com/ipfs/go-cid"
+	config "github.com/ipfs/go-ipfs-config"
+	"github.com/ipfs/go-ipfs/core"
+	ipld "github.com/ipfs/go-ipld-format"
+	pinclient "github.com/ipfs/go-pinning-service-http-client"
+	"github.com/libp2p/go-libp2p-core/host"
+	peer "github.com/libp2p/go-libp2p-core/peer"
+)
+
+type lastPin struct {
+	Time          time.Time
+	ServiceName   string
+	ServiceConfig config.RemotePinningService
+	CID           cid.Cid
+}
+
+const daemonConfigPollInterval = time.Minute / 2
+const defaultRepinInterval = 5 * time.Minute
+
+type pinMFSContext interface {
+	Context() context.Context
+	GetConfigNoCache() (*config.Config, error)
+}
+
+type pinMFSNode interface {
+	RootNode() (ipld.Node, error)
+	Identity() peer.ID
+	PeerHost() host.Host
+}
+
+type ipfsPinMFSNode struct {
+	node *core.IpfsNode
+}
+
+func (x *ipfsPinMFSNode) RootNode() (ipld.Node, error) {
+	return x.node.FilesRoot.GetDirectory().GetNode()
+}
+
+func (x *ipfsPinMFSNode) Identity() peer.ID {
+	return x.node.Identity
+}
+
+func (x *ipfsPinMFSNode) PeerHost() host.Host {
+	return x.node.PeerHost
+}
+
+func pinMFSOnChange(configPollInterval time.Duration, cctx pinMFSContext, node pinMFSNode, errCh chan<- error) error {
+	go func() {
```

Comment:
This goroutine seems to be the only thing we do. Any reason not to have the function be synchronous and then call `go pinMFSOnChange`?

Also, this function seems pretty long/difficult to follow, perhaps we can separate out the timer + state polling logic from the update logic or separate out parts of the update logic. This may also help reduce the number of channels for us to consider at any given point.